src/constraint_aware_append.c

```diff
@@ -360,13 +373,80 @@ constraint_aware_append_plan_create(PlannerInfo *root, RelOptInfo *rel, struct C
 	CustomScan *cscan = makeNode(CustomScan);
 	Plan *subplan = linitial(custom_plans);
 	RangeTblEntry *rte = planner_rt_fetch(rel->relid, root);
+	List *chunk_ris = NIL;
+	List *children = NIL;
+	ListCell *lc_child;
 
 	cscan->scan.scanrelid = 0;			 /* Not a real relation we are scanning */
 	cscan->scan.plan.targetlist = tlist; /* Target list we expect as output */
 	cscan->custom_plans = custom_plans;
-	cscan->custom_private = list_make3(list_make1_oid(rte->relid),
-									   list_copy(root->append_rel_list),
-									   list_copy(clauses));
+
+	/*
+	 * create per chunk RestrictInfo
+	 *
+	 * We also need to walk the expression trees of the restriction clauses and
+	 * update any Vars that reference the main table to instead reference the child
+	 * table (chunk) we want to exclude.
+	 */
+	switch (nodeTag(linitial(custom_plans)))
+	{
+		case T_MergeAppend:
+			children = castNode(MergeAppend, linitial(custom_plans))->mergeplans;
+			break;
+		case T_Append:
+			children = castNode(Append, linitial(custom_plans))->appendplans;
+			break;
+		default:
+			elog(ERROR, "invalid plan %d", nodeTag(linitial(custom_plans)));
+			break;
+	}
+
+	/*
+	 * we only iterate over the child chunks of this node
+	 * so the list of metadata exactly matches the list of
+	 * child nodes in the executor
+	 */
+	foreach (lc_child, children)
+	{
+		Plan *plan = get_plans_for_exclusion(lfirst(lc_child));
+
+		switch (nodeTag(plan))
+		{
+			case T_SeqScan:
+			case T_SampleScan:
+			case T_IndexScan:
+			case T_IndexOnlyScan:
+			case T_BitmapIndexScan:
+			case T_BitmapHeapScan:
+			case T_TidScan:
+			case T_SubqueryScan:
+			case T_FunctionScan:
+			case T_ValuesScan:
+			case T_CteScan:
+			case T_WorkTableScan:
+			case T_ForeignScan:
+			case T_CustomScan:
+			{
+				List *chunk_ri = NIL;
```

Comment:
maybe name chunk_clauses? Not really an ri.

Reply:
changed